pkg/services/ngalert/schedule/registry.go

```diff
@@ -233,3 +233,121 @@ func (r *alertRulesRegistry) getDiff(rules map[models.AlertRuleKey]*models.Alert
 	}
 	return result
 }
+
+type fingerprint uint64
+
+func (f fingerprint) String() string {
+	return fmt.Sprintf("%016x", uint64(f))
+}
+
+// fingerprintSeparator used during calculation of fingerprint to separate different fields. Contains a byte sequence that cannot happen in UTF-8 strings.
+var fingerprintSeparator = []byte{255}
+
+type ruleWithFolder struct {
+	rule        *models.AlertRule
+	folderTitle string
+}
+
+// fingerprint calculates a fingerprint that includes all fields except rule's Version and Update timestamp.
+func (r ruleWithFolder) Fingerprint() fingerprint {
+	rule := r.rule
+
+	sum := fnv.New64()
+
+	writeBytes := func(b []byte) {
+		_, _ = sum.Write(b)
+		_, _ = sum.Write(fingerprintSeparator)
+	}
+	writeString := func(s string) {
+		// avoid allocation when converting string to byte slice
+		writeBytes(*(*[]byte)(unsafe.Pointer(&s))) //nolint:gosec
```

Comment:
Does this needs bounds checks? I found this https://groups.google.com/g/golang-nuts/c/Zsfk-VMd_fU/m/O1ru4fO-BgAJ?pli=1

Reply:
I honestly have no idea why those boundaries are checked there (its 65535 less than Int32.Max), and the author did not provide explanations. Although I do not it is physically possible to submit such a large string via database, I ran a quick test and confirmed that the logic works even with strings whose length is int32.max. 
The thread is very interesting, but I think the solution explained there was for a general case when: 1. a string could be GCed while it is used as byte slice, and 2. to correctly specify capacity of the slice. This is not our case because: we hold a reference to the alert rule, and the lifetime of the byte slice is very short (just a single iteration), and we do not use capacity of the string.

Reply:
I left TODO to convert it to usafe.StringData once we switch to 1.20